common/replication/block_fetcher.go

```diff
@@ -475,6 +476,121 @@ func (bf *BlockFetcher) setup() {
+
+	/* If the `VerifyBlock` is not set, BlockFetcher tries to pull genesis block and intialize the `VerifyBlock` function
+	based on it. It will try to pull genesis block(block number 0) till it suceeds.*/
+	if bf.VerifyBlock == nil {
+		bf.Logger.Debugf("VerifyBlock Function is not set.")
+		bf.setupVerifierInProgress = true
+		for bf.setupVerifierInProgress {
+			// keep polling till we get a genesis block.
+			ok := bf.setUpVerifierFromGenesisBlock()
+			bf.Logger.Debugf("Try to pull genesis block %v", ok)
+			if ok {
+				bf.Logger.Debugf("Successfully pulled a genesis block.")
+				bf.setupVerifierInProgress = false
+			}
+		}
+	}
+}
+
+// setUpVerifierFromGenesisBlock connects to orderers and tries to pull genesis block from f+1 distinct
+// orderers and sets up verifier
+func (bf *BlockFetcher) setUpVerifierFromGenesisBlock() bool {
+	candidates := bf.blockSourceCandidates()
+
+	var wg sync.WaitGroup
+	wg.Add(len(candidates))
+	var blockSources []BlockSource
+	var lock sync.Mutex
+	var blocks []*common.Block
+	var sourceError error
+
+	for i := 0; i < len(candidates); i++ {
+		go func(candidate EndpointCriteria) {
+			defer wg.Done()
+			config := bf.FetcherConfig
+			config.Endpoints = []EndpointCriteria{candidate}
+			blockSource, err := bf.BlockSourceFactory(config, bf.LastConfigBlock)
+			if err != nil {
+				bf.Logger.Errorf("Failed to create block source: %v", err)
+				lock.Lock()
+				// update sourceError only if not already updated.
+				if sourceError == nil {
+					sourceError = err
+				}
+				lock.Unlock()
+				return
+			}
+			defer blockSource.Close()
+
+			lock.Lock()
+			blockSources = append(blockSources, blockSource)
+			lock.Unlock()
+
+			block := blockSource.PullBlock(0)
+			lock.Lock()
+			blocks = append(blocks, block)
+			lock.Unlock()
+
+			blockSource.Close()
+		}(candidates[i])
+	}
+	wg.Wait()
+
+	blocksEqual := func(a *common.Block, b *common.Block) bool {
+		// compare Headers
+		if a.Header.Number != b.Header.Number {
+			return false
+		}
+		if !bytes.Equal(a.Header.DataHash, b.Header.DataHash) {
+			return false
+		}
+		return true
+	}
+
+	quorum := bf.MaxByzantineNodes + 1
+	// pull f+1 blocks and compare headers
+	uniqueBlock := func(blocks []*common.Block, quorum int, compareBlocks func(a *common.Block, b *common.Block) bool) *common.Block {
+		counter := make([]int, len(blocks))
+		var selectedBlock *common.Block
+		// make sure we have f+1 identical unique blocks
+		for i, j := range blocks {
+			for _, n := range blocks {
+				if n != nil && j != nil {
+					if blocksEqual(j, n) {
+						counter[i]++
+						if counter[i] >= quorum {
+							selectedBlock = j
+							goto finish
+						}
+					}
+				}
+			}
+		}
+	finish:
+		return selectedBlock
+	}
+
+	// pull f+1 blocks and compare headers
+	genesisBlock := uniqueBlock(blocks, quorum, blocksEqual)
+	if genesisBlock == nil {
+		bf.Logger.Debugf("Couldn't get enough genesis blocks")
+		return false
+	}
+	// verify data hash on the genesis block
```

Comment:
let's verify the block data hash matches the header's hash right when we pull it, not here.

Reply:
```
		go func(candidate EndpointCriteria) {
			defer wg.Done()
			config := bf.FetcherConfig
			config.Endpoints = []EndpointCriteria{candidate}
			blockSource, err := bf.BlockSourceFactory(config, bf.LastConfigBlock)
			if err != nil {
				bf.Logger.Errorf("Failed to create block source: %v", err)
				return
			}
			defer blockSource.Close()

			lock.Lock()
			blockSources = append(blockSources, blockSource)
			lock.Unlock()

			block := blockSource.PullBlock(0)
                    dataHash := protoutil.BlockDataHash(genesisBlock.Data)
                    if  !bytes.Equal(dataHash, genesisBlock.Header.DataHash) {
                            bf.Logger.Warnf("Got a hash mismatch from %s, (exepcted %s but got %s)", candidate.Endpoint, base64.StdEncoding.EncodingString(dataHash), base64.StdEncoding.EncodingString(genesisBlock.Header.DataHash))
                          return
                      }


			lock.Lock()
			blocks = append(blocks, block)
			lock.Unlock()


		}(candidates[i])
```

Reply:
sure